### PR TITLE
SB-791 Reimplement `ArtifactDirectoryOperation` using `RepositoryPath` abstraction

### PR DESCRIPTION
**Pull request by sbespalov**

### CODE REVIEW

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryFileSystemProvider.java [hunk 1]
```diff
@@ -42,20 +42,31 @@
...
-
+    private boolean allowsForceDelete;
+    
     public RepositoryFileSystemProvider(FileSystemProvider storageFileSystemProvider)
     {
         super();
         this.storageFileSystemProvider = storageFileSystemProvider;
     }
 
+    public boolean isAllowsForceDelete()
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryFileSystemProvider.java]: Could you please rename this to `allowsForceDelete()`? Didn't we already sort this out once...?

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryFileSystemProvider.java [hunk 2]
```diff
@@ -155,10 +166,14 @@ public void createDirectory(Path dir,
                                 FileAttribute<?>... attrs)
             throws IOException
     {
-        storageFileSystemProvider.createDirectory(getTargetPath(dir), attrs);
+        storageFileSystemProvider.createDirectory(unwrap(dir), attrs);
     }
 
-    public void delete(Path path)
+    public void delete(Path path) throws IOException{
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryFileSystemProvider.java]: Cosmetic: Brace on new line, please.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java [hunk 3]
```diff
@@ -104,180 +88,19 @@ public Storage getStorage(String storageId)
         return configurationManager.getConfiguration().getStorage(storageId);
     }
 
-    @Override
-    public ArtifactInputStream getInputStream(String storageId,
-                                              String repositoryId,
-                                              String path)
-            throws IOException,
-                   NoSuchAlgorithmException
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java]: @carlspring pleace notice that `getInputStream/getOutputStream` mehtods was removed from `LayoutProvider` interface. You should use `Files.newInputStream/Files.newOutputStream` now instead.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/ArtifactDirectoryLocator.java [hunk 4]
```diff
@@ -22,51 +20,43 @@
...
 
         logger.debug("Executed (cache: " + operation.getVisitedRootPaths().size() + ")" +
-                     " visits in " + (endTime - startTime) + " ms.");
+                " visits in " + (endTime - startTime) + " ms.");
 
         getOperation().getVisitedRootPaths().clear();
     }
 
-    public String getStartingPath()
+    public RepositoryPath getStartingPath()
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/ArtifactDirectoryLocator.java]: @carlspring can you check that the original idea of this method is not distorted.

Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/ArtifactDirectoryLocator.java]: @sbespalov,

From what I've seen, the behaviour looks as expected. Thanks!

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/AbstractArtifactLocationHandler.java [hunk 5]
```diff
@@ -19,42 +24,47 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(AbstractArtifactLocationHandler.class);
+    
     private Storage storage;
-
-    private Repository repository;
-
-    private FilenameFilter filter;
-
-    private LinkedHashMap<String, List<File>> visitedRootPaths = new LinkedHashMap<>();
+    private LinkedHashMap<RepositoryPath, List<RepositoryPath>> visitedRootPaths = new LinkedHashMap<>();
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/AbstractArtifactLocationHandler.java]: Cosmetic: Could you please, leave empty lines:
* Between field declarations
* Before and after logging blocks and `if`-statements and loops
* Before `return` statements
* At the end of files

Thanks! :)

**Comment by sbespalov** [comment]
https://github.com/strongbox/strongbox/pull/262

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryFileSystemProvider.java (hunk 1 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryFileSystemProvider.java]: are you sure about that? this is common naming convention for `boolean` `getters/setters`.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryFileSystemProvider.java (hunk 2 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryFileSystemProvider.java]: done

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/ArtifactDirectoryLocator.java (hunk 4 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/ArtifactDirectoryLocator.java]: ok, thanks

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/AbstractArtifactLocationHandler.java (hunk 5 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/AbstractArtifactLocationHandler.java]: done here
will try to follow this style in the future

**Pull request closed without merging**